qemu/tests/vioser_in_use.py

```
@@ -0,0 +1,120 @@
+
+
+@error_context.context_aware
+def run(test, params, env):
+    """
+    Driver in use test:
+    1) boot guest with serial device.
+    2) Transfer data between host and guest via serialport.
+    3) run interrupt test during background stress test.
+
+    :param test: QEMU test object
+    :param params: Dictionary with the test parameters
+    :param env: Dictionary with test environment.
+    """
+
+    def run_bg_test():
+        """
+        Run serial data transfer backgroud test.
+
+        :return: return the background case thread if it's successful;
+                 else raise error.
+        """
+        error_context.context("Run serial transfer test in background",
+                              logging.info)
+        stress_thread = utils_misc.InterruptedThread(
+                virtio_serial_file_transfer.transfer_data, (params, vm),
+                {"sender": sender})
+        stress_thread.start()
+
+        check_bg_timeout = float(params.get('check_bg_timeout', 120))
+        if not utils_misc.wait_for(lambda: driver_in_use.check_bg_running(vm,
+                                   params), check_bg_timeout, 0, 1):
+            test.fail("Backgroud test is not alive!")
+        return stress_thread
+
+    driver = params["driver_name"]
+    sender = params['file_sender']
+    timeout = int(params.get("login_timeout", 360))
+    suppress_exception = params.get("suppress_exception", "no") == "yes"
+
+    vm = env.get_vm(params["main_vm"])
+    session = vm.wait_for_login()
+    if params["os_type"] == "windows":
+        session = utils_test.qemu.windrv_check_running_verifier(session, vm,
+                                                                test, driver,
+                                                                timeout)
+
+    bg_thread = run_bg_test()
+    globals().get(params["interrupt_test"])(test, params, vm, session)
+    if bg_thread:
+        bg_thread.join(timeout=timeout, suppress_exception=suppress_exception)
```

You might get return value of the thread and check result of it.

---

guest will be reboot/shutdown/xxx, the result is error, our step is to ignore this error because it is expected result.

---

@lijinlijin Okay, make sense